pkg/channel/jetstream/dispatcher/consumer.go

```diff
@@ -73,51 +73,27 @@ func (c *Consumer) Close() error {
 }
 
 func (c *Consumer) MsgHandler(msg *nats.Msg) {
-	logger := c.logger.With(zap.String("msg_id", msg.Header.Get(nats.MsgIdHdr)))
-	ctx := logging.WithLogger(c.ctx, logger)
-	tickerCtx, tickerCancel := context.WithCancel(c.ctx)
-
-	tickerDone := make(chan struct{})
-
-	go func() {
-		defer close(tickerDone)
-
-		// TODO(dan-j): this should be a fraction of the Consumer's AckWait
-		ticker := time.NewTicker(10 * time.Second)
-		defer ticker.Stop()
-
-		for {
-			select {
-			case <-tickerCtx.Done():
-				return
-			case <-ticker.C:
-				if err := msg.InProgress(nats.Context(tickerCtx)); err != nil && !errors.Is(err, context.Canceled) {
-					logging.FromContext(ctx).Errorw("failed to mark message as in progress", zap.Error(err))
-				}
-			}
-		}
-	}()
-
 	go func() {
+		logger := c.logger.With(zap.String("msg_id", msg.Header.Get(nats.MsgIdHdr)))
+		ctx := logging.WithLogger(c.ctx, logger)
 		var result protocol.Result
 
-		// wrap the handler in a local function so that the tickerCtx is cancelled even if a panic occurs.
-		func() {
-			defer tickerCancel()
-			result = c.doHandle(ctx, msg)
-		}()
-
-		// wait for the ticker to stop to prevent attempts to mark the message as in progress after it has been acked
-		// or nacked
-		<-tickerDone
+		result = c.doHandle(ctx, msg)
```

Comment:
This is looking really good now, my only concern is how we cancel requests when the AckWait expires, as this was the main problem for us which was solved by the ticker goroutine.

The messages are buffered in the client before calling our `MsgHandler` method, so simply creating a `context.WithTimeout(ctx, config.AckWait)` may actually result in a message being handled more than once. I was hoping the `msg.Metadata().Timestamp` would be the time from which the ack-wait starts from, but this is the time the message was added to the stream, and doesn't change with redeliveries.

I think you might have to do something like:

```go
meta, _ := msg.Metadata()
// probably store this as a struct field on the Consumer
info, _ := c.jsSub.ConsumerInfo()

// if each delivery has timed out, then multiplying the number of deliveries by the ack wait will give us the
// duration from publish which this attempt will be ack-waited
ackDurationFromPublish := time.Duration(meta.NumDelivered) * info.Config.AckWait
// add a bit of jitter
jitter := 500 * time.Millisecond

// the deadline is the published timestamp plus our duration calculated above
deadline := meta.Timestamp.Add(ackDurationFromPublish).Add(-jitter)

// if previous deliveries were explicitly nacked earlier than the deadline, then our actual deadline will be earlier
// than the deadline above
ackDeadlineFromNow := time.Now().Add(info.Config.AckWait).Add(-jitter)
if deadline.After(ackDeadlineFromNow) {
	deadline = ackDeadlineFromNow
}

ctx, cancel := context.WithDeadline(ctx, deadline)
defer cancel()

result = c.doHandle(ctx, msg)

...
```

Reply:
The jitter should also probably be quite a bit lower, or a fraction of AckWait, because I imagine some deployments may have a small AckWait where 500ms may actually cause more problems than they solve.

Reply:
So, I've continued some more reading and now I'm at the conclusion we're building the consumers wrong. I'm happy to take this into a new issue though, if you think that's appropriate.

We're using push consumers, and reading this page, they're described as legacy: https://natsbyexample.com/examples/jetstream/push-consumer/go

I'm not sure who the author of natsbyexample.com is, but it's spot on with the issues which arise when consumers can't keep up.

We should probably refactor where consumers are created and use a pull consumer instead, using `FetchNoWait` to process as many as possible in a batch size we think is reasonable: https://natsbyexample.com/examples/jetstream/pull-consumer/go

What do you reckon @astelmashenko, address here or a new issue?

Reply:
pull consumer looks more reasonable to me, because it allows to better control back-pressure. I'm not sure about how to controll back-pressure in golang world with goroutines, how to control max inflight requests?

Regarding the comment with deadline calculation, it looks complex and I'd just use 
`ackDeadlineFromNow := time.Now().Add(info.Config.AckWait).Add(-jitter)`
I thinkg probability of duplicate message in case of timeout will be almost the same.

But I'll add all those calculations because it is already implemented.

Reply:
> This is looking really good now, my only concern is how we cancel requests when the AckWait expires, as this was the main problem for us which was solved by the ticker goroutine.

It is covered already by jitter, see defaults.go
`consumerConfig.AckWait = retryConfig.RequestTimeout + jitter`
and I calculate context timeout as just `retryConfig.RequestTimeout`
`ctx, cancel = context.WithTimeout(ctx, c.sub.RetryConfig.RequestTimeout)`

so AckWait is bigger then RequestTimeout and there is jitter time to Ack message if request almost reached timeout.
Maybe jitter should be lower, I set it 5sec, maybe just change it to 500ms.

Example:
RequestTimeout = 30s 
AckWait = 30s + 500ms

case1: request completed in 2s, it's ok message Ack'ed
case2: request timed out, context cancelled in 30s, there is 500ms to NaK message, even if msg is not NaK'ed it will be redelivered in any case
case3: request completed in 29,999s, there is 500ms to Ack message

the only concern is the last case3, IMO probability of such case is pretty low, at least because requestTimeout should be bigger then p95 of request processing time by a target service.

Should we do all those complex calculations? do you still see it reasonable?

Reply:
So you're correct, but the reason we added the ticker was to address case 3 because this was happening a lot if the subscriber is scaling up and the request gets queued in knative's activator. We were observing the message being pending on the NATS client for 10-20 seconds before our MsgHandler is called.

I see two options:

1. Move to pull consumer and do away with my obscure maths
2. Do my obscure maths and create a new issue to move to pull consumer, then we'll remove this as part of that issue